### PR TITLE
[NEW] Standard Importer Structure

### PR DESCRIPTION
**Pull request by pierre-lehnen-rc**
## Proposed changes

This PR aims to simplify the data importers process by storing the data of any importer in a standard format that can be converted by the same process.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Hotfix (a major bugfix that has to be merged asap)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [X] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [X] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat)
- [X] Lint and unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [X] I have added necessary documentation (if applicable)
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by wreiske** [comment]
This is very cool. I'd like to help drive connecting this up with another project aimed at simplifying access to third-party data providers such as Google Contacts, GSuite, ADP, BambooHR, BreezyHR, Salesforce, Workday, etc. You can read about it here https://github.com/easybread/easybread/tree/develop/packages/core#easybread and see the documentation https://github.com/easybread/easybread/wiki/GeneralConcepts

This would make it really cool for rocket.chat users that have an HR or directory like BambooHR, ADP, Gsuite, but no AD or LDAP.
With this method, administrators wouldn't have to manually create users in RC and could sync right up with their HR provider without needing LDAP or AD.

Also, easybread plans to support LDAP and AD as an adapter soon so we could potentially also take over RC's LDAP user sync code with the easybread adapter. Maybe even slack and other adapters could be implemented in easybread and then it would all just be in one module in RC and as easybread implements more adapters they instantly become available with a npm update of easybread on RC.

I have a freelance developer that has done the majority of the work on easybread. We would be happy to work with the Rocket.Chat team to build out the UI functionality and having easybread importers available as part of the setup wizard and also in the administration dashboard. Imagine setting up a new Rocket.Chat server, clicking on an import from existing directory button, and being able to select any third party provider without any need for Rocket.Chat's team to maintain the adapters. 

image

**Comment by geekgonecrazy** [comment]
i'll create a discussion on open.  I don't want to let this sit un-answered too long

**Comment by geekgonecrazy** [comment]
Any particular reason we haven’t moved forward and merged?

**Comment by pierre-lehnen-rc** [comment]
> Any particular reason we haven’t moved forward and merged?

It's missing a few things, but I'll finish it for this month's release.

File: app/importer/server/classes/ImporterBase.js [hunk 1]
```diff
@@ -196,7 +154,71 @@ export class Base {
 			throw new Error(`Channels in the selected data wasn't found, it must but at least an empty array for the ${ this.info.name } importer.`);
 		}
 
-		return this.updateProgress(ProgressStep.IMPORTING_STARTED);
+		this.updateProgress(ProgressStep.IMPORTING_STARTED);
+		this.reloadCount();
+		const started = Date.now();
+		const startedByUserId = Meteor.userId();
+		console.log(importSelection);
```
Comment (pierre-lehnen-rc) [inline app/importer/server/classes/ImporterBase.js:161]:
```suggestion
```

File: app/importer/server/definitions/IImportUser.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+export interface IImportUser {
+	// #ToDo: Remove this _id, as it isn't part of the imported data
```
Comment (pierre-lehnen-rc) [inline app/importer/server/definitions/IImportUser.ts:2]: oops

File: app/importer/server/classes/ImporterBase.js (hunk 1 above)
Comment (KevLehman) [inline app/importer/server/classes/ImporterBase.js:161]: 😬

File: app/importer/server/startup/setImportsToInvalid.js [hunk 3]
```diff
@@ -8,7 +8,7 @@ function runDrop(fn) {
 	try {
 		fn();
 	} catch (e) {
-		console.log('errror', e); // TODO: Remove
+		console.log('error', e); // TODO: Remove
```
Comment (KevLehman) [inline app/importer/server/startup/setImportsToInvalid.js:11]: I think this either should be a `logger` call or a `console.error`

File: app/importer/server/classes/ImportDataConverter.ts [hunk 4]
```diff
@@ -0,0 +1,832 @@
...
+	}
+
+	insertRoom(roomData: IImportChannel, startedByUserId: string): void {
+		// Find the rocketchatId of the user who created this channel
+		const creatorId = this.getRoomCreatorId(roomData, startedByUserId);
+		const members = this.convertImportedIdsToUsernames(roomData.users, roomData.t !== 'd' ? creatorId : undefined);
+
+		if (roomData.t === 'd') {
+			if (members.length < roomData.users.length) {
+				console.warn('One or more imported users not found: ${ roomData.users }');
```
Comment (sampaiodiego) [inline app/importer/server/classes/ImportDataConverter.ts:688]: thx to LTGM.. this log will not show what is expected =)

btw, any reason to not use Logger?

Comment (pierre-lehnen-rc) [reply app/importer/server/classes/ImportDataConverter.ts:688]: Not anymore, just didn't update this part of the code.

**Review by sampaiodiego — APPROVED** [review]

**Pull request merged**